shopfloor_base/services/service.py

```diff
@@ -0,0 +1,267 @@
+                    "data": {next_state: data or {}},
+                    "next_state": next_state,
+                }
+            )
+
+        elif data:
+            response["data"] = data
+
+        if message:
+            response["message"] = message
+
+        if popup:
+            response["popup"] = popup
+
+        return response
+
+    _requires_header_menu = False
+    _requires_header_profile = False
+
+    def _get_openapi_default_parameters(self):
+        defaults = super()._get_openapi_default_parameters()
+        # Normal users can't read an API key, ignore it using sudo() only
+        # because it's a demo key.
+        demo_api_key = self.env.ref("shopfloor.api_key_demo", raise_if_not_found=False)
+        if demo_api_key:
+            demo_api_key = demo_api_key.sudo()
+
+        service_params = [
+            {
+                "name": "API-KEY",
+                "in": "header",
+                "description": "API key for Authorization",
+                "required": True,
+                "schema": {"type": "string"},
+                "style": "simple",
+                "value": demo_api_key.key if demo_api_key else "",
+            },
+        ]
+        if self._requires_header_menu:
+            # Try to first the first menu that implements the current service.
+            # Not all usages have a process, in that case, we'll set the first
+            # menu found
+            menu = self.env["shopfloor.menu"].search(
+                [("scenario", "=", self._usage)], limit=1
+            )
+            if not menu:
+                menu = self.env["shopfloor.menu"].search([], limit=1)
+            service_params.append(
+                {
+                    "name": "SERVICE_CTX_MENU_ID",
+                    "in": "header",
+                    "description": "ID of the current menu",
+                    "required": True,
+                    "schema": {"type": "integer"},
+                    "style": "simple",
+                    "value": menu.id,
+                }
+            )
+        if self._requires_header_profile:
+            profile = self.env["shopfloor.profile"].search([], limit=1)
+            service_params.append(
+                {
+                    "name": "SERVICE_CTX_PROFILE_ID",
+                    "in": "header",
+                    "description": "ID of the current profile",
+                    "required": True,
+                    "schema": {"type": "integer"},
+                    "style": "simple",
+                    "value": profile.id,
+                }
+            ),
+        defaults.extend(service_params)
+        return defaults
+
+    @property
+    def data(self):
+        return self._actions_for("data")
+
+    @property
+    def data_detail(self):
+        return self._actions_for("data_detail")
+
+    @property
+    def msg_store(self):
+        return self._actions_for("message")
+
+    # TODO: maybe to be proposed to base_rest
```

Comment:
May be it's time ?

Reply:
well, IDK, Current solution is a bit hack-ish and w/ the new api we might be able to add this validation via decorators.
I'd keep it as it is now and get back to it on a second round.

Reply:
Ok